Add color_temp_command_template for mqtt light component

jonfink opened the pull request:
**Description:**

Home assistant internal unit for color temperature is Mired but some
systems (e.g., SmartThings through mqtt bridge) use Kelvin. Conversion
between these types can be handled with the value template
concept (used previously to unpack json values).

Added a "color_temp_command_template" configuration option for MQTT
lights that specifies a function to be applied to the HA-internal
color temperature value to get a light-compatible value.

**Related issue (if applicable):** 

n/a

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** 

home-assistant/home-assistant.github.io#1673

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
_n/a code interacts with mqtt devices_

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

jonfink committed:
Add color_temp_command_template for mqtt light component

Home assistant internal unit for color temperature is Mired but some
systems (e.g., SmartThings through mqtt bridge) use Kelvin. Conversion
between these types can be handled with the value template
concept (used previously to unpack json values).

Added a "color_temp_command_template" configuration option for MQTT
lights that specifies a function to be applied to the HA-internal
color temperature value to get a light-compatible value.

jonfink committed:
Fix long lines in light/mqtt.py

jonfink committed:
Fix logging-not-lazy pylint error in light/mqtt.py

balloob commented on homeassistant/components/light/mqtt.py:
```diff
@@ -85,7 +89,9 @@ def setup_platform(hass, config, add_devices, discovery_info=None):
             CONF_RGB: config.get(CONF_RGB_VALUE_TEMPLATE),
-            CONF_COLOR_TEMP: config.get(CONF_COLOR_TEMP_VALUE_TEMPLATE)
+            CONF_COLOR_TEMP: config.get(CONF_COLOR_TEMP_VALUE_TEMPLATE),
+            CONF_COLOR_TEMP+"_COMMAND":
```
This will become `color_temp_command_COMMAND` ?

jonfink replied:
The dictionary key will be `color_temp_command` to go alongside the existing dictionary key of `color_temp`. 

Honestly, I wasn't sure how best to handle this as it seemed more far-reaching than desired to make a change in the mqtt light component that required an addition to the homeassistant.const module. If that is the standard practice, then `CONF_COLOR_TEMP_COMMAND` could be added to `homeassistant/const.py`...

MartinHjelmare replied:
Just use `CONF_COLOR_TEMP_COMMAND` as key here.

balloob commented on homeassistant/components/light/mqtt.py:
```diff
@@ -283,7 +298,7 @@ def turn_on(self, **kwargs):
 
         if should_update:
-            self.schedule_update_ha_state()
+            self.update_ha_state()
```
please change this back.

jonfink replied:
Done in 66012e9

balloob commented on homeassistant/components/light/mqtt.py:
```diff
@@ -293,4 +308,4 @@ def turn_off(self, **kwargs):
             # Optimistically assume that switch has changed state.
             self._state = False
-            self.schedule_update_ha_state()
+            self.update_ha_state()
```
please change this back.

jonfink replied:
Done in 66012e9

jjensn commented:
What needs to be done here to get this approved? @jonfink how can I help?

jonfink committed:
Revert to use scheduled ha update functions

jonfink commented:
I think I've addressed all of the above comments with the exception of the dictionary keys used to instantiate the MqttLight class `template` argument. I'm not entirely sure on the rationale in general for these keys coming from `homeassistant.const` since they are entirely internal to the mqtt light component, but if that part is important it would be a relatively small change to add an additional value to the const module.

MartinHjelmare commented on homeassistant/components/light/mqtt.py:
```diff
@@ -266,7 +277,11 @@ def turn_on(self, **kwargs):
            self._topic[CONF_COLOR_TEMP_COMMAND_TOPIC] is not None:
-            color_temp = int(kwargs[ATTR_COLOR_TEMP])
+            color_temp = int(float((
+                self.templates[CONF_COLOR_TEMP+"_COMMAND"](
```
Use `CONF_COLOR_TEMP_COMMAND` as key here.

jjensn replied:
Hey @MartinHjelmare -- doesn't CONF_COLOR_TEMP_COMMAND have to be added to https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/const.py ? Is that the change you are calling for?

MartinHjelmare reviewed (changes requested):
See comments. There was a unit test failing also. Fix that and add test(s) for the new functionality.

jjensn commented:
I need to ask a serious question here and in no way am I trying to be snarky.

How do the collaborators feel about this approach vs using a configuration option and doing the conversion from mired to kelvin there?

I ask because, someone (me) is going to want to add RGB templating next, and you can't quite do RGB -> hue/sat very well in a template (as far as I know, the math would be too complex).

So why not just add two more variables, color_temp_mode (similar to flux component: https://home-assistant.io/components/switch.flux/), and color_mode (rgb vs xy vs hue/sat) ?

I think really, the only reason @jonfink is adding the support for this, is so we can convert HA's default color_temp (mired) to kelvin. I can't see any other use for templating out the command other than that conversion.

balloob commented:
@jjensn I like that approach 👍

homeassistant commented:
Hi @jonfink, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5105).

Once you do that we will be able to review and accept this pull request.

Thanks!

andrey-git commented:
@jonfink any updates on this?

balloob commented:
This PR seems to have gone stale. Closing it. You can reopen it when you're ready to finish it.

raccettura commented:
Any chance on resurrecting this? Would be handy to control osram bulbs via HA->ST.
